File: clients/ios/Lextures/Core/Offline/OfflineService.swift
import Foundation
import Observation

/// Shared offline coordinator: cache reads, download storage, and write outbox replay.
@MainActor
@Observable
final class OfflineService {
    static let shared = OfflineService()

    private(set) var pendingCount = 0
    private(set) var storageBytes = 0
    private(set) var outboxItems: [OutboxItem] = []
    private(set) var metrics = OfflineSyncMetrics()

    private var ownerKey = "anonymous"
    private var cacheStore = CacheStore(ownerKey: "anonymous")
    private var outboxStore = OutboxStore(ownerKey: "anonymous")
    private var downloadStore = DownloadStore(ownerKey: "anonymous")
    private var syncEngine: SyncEngine

    private init() {
        let outbox = OutboxStore(ownerKey: "anonymous")
        outboxStore = outbox
        syncEngine = SyncEngine(outbox: outbox, ownerKey: "anonymous")
    }

    func configure(accessToken: String?) {
        let key = NotebookStore.jwtSubject(from: accessToken) ?? "anonymous"
        if key != ownerKey {
            ownerKey = key
            cacheStore = CacheStore(ownerKey: key)
            outboxStore = OutboxStore(ownerKey: key)
            downloadStore = DownloadStore(ownerKey: key)
            syncEngine = SyncEngine(outbox: outboxStore, ownerKey: key)
        }
        Task { await refreshState() }
    }

    func clearAllOnLogout() {
        Task {
            await cacheStore.clearAll()
            await downloadStore.clearAll()
            await outboxStore.clearAll()
            await refreshState()
        }
    }

    /// Cache-first fetch: returns cached data immediately; refreshes when online.
    func cachedFetch<T: Codable>(
        key: String,
        accessToken: String,
        fetch: () async throws -> T
    ) async throws -> (value: T, cached: Cached<T>?) {
        let online = NetworkMonitor.shared.isOnline
        let existing = await cacheStore.get(T.self, key: key)

        if online {
            do {
                let fresh = try await fetch()
                try await cacheStore.put(fresh, key: key)
                await refreshState()
                return (fresh, Cached(value: fresh, fetchedAt: Date()))
            } catch {
                if let existing {
                    return (existing.value, existing)
                }
                throw error
            }
        }

        if let existing {
            return (existing.value, existing)
        }
        throw APIError.transport(URLError(.notConnectedToInternet))
    }

    /// Enqueue a mutation when offline (or when caller prefers queued replay).
    func enqueueMutation(
        method: String,
        path: String,
        body: (any Encodable)?,
        label: String,
        accessToken: String?,
        preferQueue: Bool = false,
        idempotencyKey: String? = nil
    ) async throws -> OutboxItem {
        let bodyJSON: String?
        if let body {
            let data = try JSONEncoder().encode(AnyEncodable(body))
            bodyJSON = String(data: data, encoding: .utf8)
        } else {
            bodyJSON = nil
        }

        let online = NetworkMonitor.shared.isOnline
        if online && !preferQueue, let accessToken {
            let bodyData = bodyJSON.flatMap { $0.data(using: .utf8) }
            let resolvedKey = idempotencyKey ?? UUID().uuidString.lowercased()
            do {
                _ = try await APIClient().requestRaw(
                    path: path,
                    method: method,
                    bodyData: bodyData,
                    authorized: true,
                    accessToken: accessToken,
                    idempotencyKey: resolvedKey
                )
                await outboxStore.markApplied(idempotencyKey: resolvedKey)
                await refreshState()
                return OutboxItem(
                    id: resolvedKey,
                    createdAt: Date(),
                    sequence: -1,
                    method: method,
                    path: path,
                    bodyJSON: bodyJSON,
                    label: label,
                    status: .synced,
                    lastError: nil
                )
            } catch let error as APIError {
                if case .transport = error {
                    // Fall through to outbox enqueue.
                } else {
                    throw error
                }
            }
        }

        let item = await outboxStore.enqueue(
            method: method,
            path: path,
            bodyJSON: bodyJSON,
            label: label,
            id: idempotencyKey
        )
        await refreshState()
        return item
    }

    func downloadContent(key: String, data: Data, fileName: String, mimeType: String?) async throws {
        try await downloadStore.save(key: key, data: data, fileName: fileName, mimeType: mimeType)
        await refreshState()
    }

    func downloadedData(key: String) async -> Data? {
        await downloadStore.loadData(key: key)
    }

    func isDownloaded(key: String) async -> Bool {
        await downloadStore.isDownloaded(key: key)
    }

    func syncNow(accessToken: String?) async {
        guard let accessToken, NetworkMonitor.shared.isOnline else { return }
        metrics = await syncEngine.sync(accessToken: accessToken)
        await refreshState()
    }

    func retryOutboxItem(id: String, accessToken: String?) async {
        await outboxStore.retry(id: id)
        await syncNow(accessToken: accessToken)
    }

    func refreshState() async {
        pendingCount = await outboxStore.pendingCount()
        outboxItems = await outboxStore.allItems()
        let cacheBytes = await cacheStore.totalSizeBytes()
        let downloadBytes = await downloadStore.totalSizeBytes()
        storageBytes = cacheBytes + downloadBytes
        metrics = await syncEngine.currentMetrics()
    }

    func clearStorage() async {
        await cacheStore.clearAll()
        await downloadStore.clearAll()
        await refreshState()
    }
}

private struct AnyEncodable: Encodable {
    private let encode: (Encoder) throws -> Void

    init(_ value: any Encodable) {
        encode = value.encode
    }

    func encode(to encoder: Encoder) throws {
        try encode(encoder)
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/offline/OfflineService.kt
package com.lextures.android.core.offline

import android.content.Context
import com.lextures.android.core.lms.CourseSummary
import com.lextures.android.core.network.ApiClient
import com.lextures.android.core.network.ApiError
import com.lextures.android.core.notebook.NotebookStore
import java.util.UUID
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.KSerializer
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

class OfflineService private constructor(context: Context) {
    private val appContext = context.applicationContext
    private val json = Json { ignoreUnknownKeys = true }
    private val apiClient = ApiClient()

    private var ownerKey = "anonymous"
    private lateinit var cacheStore: CacheStore
    private lateinit var outboxStore: OutboxStore
    private lateinit var downloadStore: DownloadStore
    private lateinit var syncWorker: SyncWorker
    val networkMonitor = NetworkMonitor(appContext)

    private val _pendingCount = MutableStateFlow(0)
    val pendingCount: StateFlow<Int> = _pendingCount.asStateFlow()

    private val _storageBytes = MutableStateFlow(0L)
    val storageBytes: StateFlow<Long> = _storageBytes.asStateFlow()

    private val _outboxItems = MutableStateFlow<List<OutboxItem>>(emptyList())
    val outboxItems: StateFlow<List<OutboxItem>> = _outboxItems.asStateFlow()

    init {
        rebuildStores("anonymous")
    }

    fun configure(accessToken: String?) {
        val key = NotebookStore.jwtSubject(accessToken) ?: "anonymous"
        if (key != ownerKey) {
            ownerKey = key
            rebuildStores(key)
        }
        refreshState()
    }

    fun clearAllOnLogout() {
        cacheStore.clearAll()
        downloadStore.clearAll()
        outboxStore.clearAll()
        refreshState()
    }

    suspend fun <T> cachedFetch(
        key: String,
        accessToken: String,
        serializer: KSerializer<T>,
        fetch: suspend () -> T,
    ): Pair<T, Cached<T>?> {
        val online = networkMonitor.isOnline.value
        val existing = cacheStore.get(key, serializer)

        if (online) {
            return try {
                val fresh = fetch()
                cacheStore.put(key, fresh, serializer)
                refreshState()
                fresh to Cached(fresh, java.time.Instant.now())
            } catch (e: Exception) {
                if (existing != null) {
                    existing.value to existing
                } else {
                    throw e
                }
            }
        }

        if (existing != null) {
            return existing.value to existing
        }
        throw ApiError.Transport(java.io.IOException("Device is offline"))
    }

    suspend fun enqueueMutation(
        method: String,
        path: String,
        bodyJson: String?,
        label: String,
        accessToken: String?,
        preferQueue: Boolean = false,
        idempotencyKey: String? = null,
    ): OutboxItem {
        val online = networkMonitor.isOnline.value
        if (online && !preferQueue && !accessToken.isNullOrBlank()) {
            val resolvedKey = idempotencyKey ?: UUID.randomUUID().toString()
            try {
                apiClient.requestRaw(
                    path = path,
                    method = method,
                    body = bodyJson,
                    accessToken = accessToken,
                    idempotencyKey = resolvedKey,
                )
                outboxStore.markApplied(resolvedKey)
                refreshState()
                return OutboxItem(
                    id = resolvedKey,
                    method = method,
                    path = path,
                    bodyJson = bodyJson,
                    label = label,
                    status = OutboxStatus.Synced.name,
                )
            } catch (e: ApiError.Transport) {
                // Fall through to outbox.
            } catch (e: Exception) {
                throw e
            }
        }

        val item = outboxStore.enqueue(method, path, bodyJson, label, idempotencyKey)
        refreshState()
        return item
    }

    suspend fun syncNow(accessToken: String?) {
        if (accessToken.isNullOrBlank() || !networkMonitor.isOnline.value) return
        syncWorker.sync(accessToken)
        refreshState()
    }

    suspend fun retryOutboxItem(id: String, accessToken: String?) {
        outboxStore.retry(id)
        syncNow(accessToken)
    }

    fun refreshState() {
        _pendingCount.value = outboxStore.pendingCount()
        _outboxItems.value = outboxStore.allItems()
        _storageBytes.value = cacheStore.totalSizeBytes() + downloadStore.totalSizeBytes()
    }

    fun clearStorage() {
        cacheStore.clearAll()
        downloadStore.clearAll()
        refreshState()
    }

    fun downloadContent(key: String, data: ByteArray, fileName: String, mimeType: String?) {
        downloadStore.save(key, data, fileName, mimeType)
        refreshState()
    }

    fun downloadedData(key: String): ByteArray? = downloadStore.loadData(key)

    fun isDownloaded(key: String): Boolean = downloadStore.isDownloaded(key)

    private fun rebuildStores(key: String) {
        cacheStore = CacheStore(appContext, key)
        outboxStore = OutboxStore(appContext, key)
        downloadStore = DownloadStore(appContext, key)
        syncWorker = SyncWorker(appContext, outboxStore, key, apiClient)
    }

    companion object {
        @Volatile
        private var instance: OfflineService? = null

        fun get(context: Context): OfflineService =
            instance ?: synchronized(this) {
                instance ?: OfflineService(context.applicationContext).also { instance = it }
            }
    }
}

/** Convenience for course list cached fetch. */
suspend fun OfflineService.fetchCoursesCached(
    accessToken: String,
    fetch: suspend () -> List<CourseSummary>,
): Pair<List<CourseSummary>, Cached<List<CourseSummary>>?> =
    cachedFetch(
        key = OfflineCacheKey.courses(),
        accessToken = accessToken,
        serializer = kotlinx.serialization.builtins.ListSerializer(CourseSummary.serializer()),
        fetch = fetch,
    )
